providers/gitlab.go

```diff
@@ -213,8 +256,37 @@ func (p *GitLabProvider) EnrichSessionState(ctx context.Context, s *sessions.Ses
 		return fmt.Errorf("group membership check failed: %v", err)
 	}
 
+	p.addProjectMembershipToState(ctx, s)
+
 	s.User = userInfo.Username
 	s.Email = userInfo.Email
 
 	return nil
+
+}
+
+// addProjectMembership adds projects into session.Groups
+func (p *GitLabProvider) addProjectMembershipToState(ctx context.Context, s *sessions.SessionState) {
+	if len(p.Projects) == 0 {
+		return
+	}
+
+	// Iterate over projects, check if oauth2-proxy can get project information on behalf of the user
+	for _, project := range p.Projects {
```

Comment:
I think this is fine (as opposed to querying for all projects like you mentioned).

There's the security benefit that we only expose the projects that we care about in the oauth2-proxy configuration in the session.Groups. It doesn't lead to a potential data leak?

Maybe this is less of an issue for project.  This same concept is in Google Groups, it only queries for ones Oauth2-Proxy knows about.  It doesn't just pull all groups a user is in potentially pulling some sensitive things it shouldn't

Reply:
You could also parallelize these requests if performance is a concern in large project list scenarios?

Reply:
The two arguments were in my mind. For sure it's a bit out of scope for oauth2-proxy to get all the repos from a user. My questioning was more about reducing overhead on the Gitlab API entry point. But it's a nitpick and the current implementation is okay. It don't think parallelize is necessary here.